### PR TITLE
[inductor] add C-shim for index_put

### PR DESCRIPTION
**Pull request by ColinPeppler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116667

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -4280,11 +4280,28 @@ def codegen(self, wrapper):
...
+
+        if (
+            V.graph.aot_mode
+            and V.graph.cpp_wrapper
+            and config.aot_inductor.abi_compatible
+        ):
+            # Make the fallback call ABI-compatible in the C++ wrapper file.
+            kernel = kernel.replace("at::", "aoti_torch_")
+            indices_str = (
+                f"std::vector<AtenTensorHandle>{{{', '.join(indices)}}}.data()"
```
Comment (aakhundov) [inline torch/_inductor/ir.py:4295]: Not sure if `std` is allowed in the ABI-compatible mode or whether it should be pure C. cc @desertfire.

**Comment by ColinPeppler** [comment]
Got an implementation to work that doesn't use `std::vector`. But need to rewrite this to call `at::index_put_out`.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -2146,6 +2151,33 @@ def generate_scatter_fallback(
...
+            num_indices = str(
+                len(indices)
+            )  # num_indices for indexing into indices array
+            tensor_handle_array_var = (
+                f"tensor_handle_array_{next(self.kernel_callsite_id)}"
+            )
+            self.writeline(
+                f"AtenTensorHandle {tensor_handle_array_var}[] = {{{', '.join(indices)}}};"
+            )
+            args = [x, tensor_handle_array_var, values, accumulate, num_indices]
```
Comment (ColinPeppler) [inline torch/_inductor/codegen/wrapper.py:2171]: example codegen
```
AtenTensorHandle tensor_handle_array_1[] = {arg1_1};
aoti_torch_index_put_out(buf0, buf0, tensor_handle_array_1, arg2_1, 1, 1);
```

File: torch/csrc/inductor/aoti_torch/shim_common.cpp [hunk 3]
```diff
@@ -638,6 +639,27 @@ AOTITorchError aoti_torch_scatter_reduce_out(
   });
 }
 
+AOTITorchError aoti_torch_index_put_out(
+    AtenTensorHandle out,
+    AtenTensorHandle self,
+    const AtenTensorHandle* indices,
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_torch/shim_common.cpp:645]: In other C-shim APIs, the array length comes right after the pointer, e.g. https://github.com/pytorch/pytorch/blob/0ef1266bc631fa8e91bf26a052eacf4b7173ff13/torch/csrc/inductor/aoti_torch/c/shim.h#L256 . Let's keep it that way here.

**Review by desertfire — CHANGES_REQUESTED** [review]
LGTM other than the parameter order comment I left.

**Review by desertfire — APPROVED** [review]

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Pull request closed without merging**